### PR TITLE
Add PDF extraction skill and improve error handling

### PR DESCRIPTION
**Pull request by Huyuochi**
目的

  新增 pdf-extract 内置 skill，让模型遇到用户提供的本地 PDF 时可以离线提取文本，扫描件则回退为分页
  PNG，行为与 openclaw 内置的 pdf tool 逻辑对齐。

  实现

  - 算法移植：builtin-skills/pdf-extract/scripts/extract.mjs 两阶段处理 — 先用 pdfjs-dist
  抽文本，文本量低于阈值则用 @napi-rs/canvas 把每页渲染成 PNG。限制（5 MiB / 4 页 / 200 000 字符 / 4
  000 000 像素）与 gateway 默认值保持一致。
  - 输出契约：stdout 单行 JSON，stderr 仅保留硬错误；显式传 verbosity: 0 屏蔽 pdfjs 的“Indexing all
  PDF objects”等告警，保证成功路径 stderr 为空。
  - workspace 依赖解析：skill 首次启动会被复制到 ~/.openclaw/workspace/skills/，失去 gateway 的
  node_modules 祖先关系。extract.mjs 改用 OPENCLAW_INSTALL_ROOT 通过 createRequire 从 gateway/ 或
  gateway.asar 复用已打包的 pdfjs-dist 与 @napi-rs/canvas，本地（gateway 直接调用）路径不变。
  - Windows node 代理：scripts/afterPack.js 在 runtime/ 下写入 node.cmd，透传 %* 到 <Product>
  Helper.exe 并设 ELECTRON_RUN_AS_NODE=1，与 macOS 的 runtime/node shell 代理对称 — 否则 Windows
  打包后 node.exe 被替换为 Electron，skill 无法用 node 命令启动。

  主要改动

  - builtin-skills/pdf-extract/SKILL.md、examples/、reference/：skill
  描述、调用示例、输出格式与算法文档。
  - builtin-skills/pdf-extract/scripts/extract.mjs：提取主脚本（172 行），含 OPENCLAW_INSTALL_ROOT
  回退解析。
  - scripts/afterPack.js：新增 buildWindowsNodeProxyScript 与 node.cmd 写入逻辑。
  - scripts/afterPack.test.js：回归用例，校验 node.cmd 写入、%* 透传、不绑定 CLI 入口。

### CODE REVIEW

File: builtin-skills/pdf-extract/scripts/extract.mjs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+function fail(msg, code = 1) {
+  process.stderr.write(`pdf-extract: ${msg}\n`);
+  process.exit(code);
+}
+
+function bundledOpenclawPackageJsons() {
+  const installRoot = process.env.OPENCLAW_INSTALL_ROOT;
+  if (!installRoot) return [];
+  return ["gateway", "gateway.asar"].map((gatewayRoot) =>
+    path.join(installRoot, gatewayRoot, "node_modules", "openclaw", "package.json"));
```
Comment (oriengy) [inline builtin-skills/pdf-extract/scripts/extract.mjs:51]: **`gateway.asar` is a file, not a directory — `existsSync` will return false outside an Electron process.**

In ASAR-mode production builds, the only artefact under `OPENCLAW_INSTALL_ROOT` is `gateway.asar` (no `gateway/`). The fallback path:

```js
path.join(installRoot, "gateway.asar", "node_modules", "openclaw", "package.json")
```

…only resolves to a real, readable file when fs is patched to descend into asar archives, i.e. when the runtime is Electron with `ELECTRON_RUN_AS_NODE=1`. With a vanilla system Node 22, `existsSync(...)` returns `false` and the candidate is silently skipped; the script then prints `pdfjs-dist not available: ...; also tried OPENCLAW_INSTALL_ROOT=...` and exits with a misleading error.

Given SKILL.md tells the agent to just run `node extract.mjs file.pdf` and the gateway-bundled Node binary has been replaced by an Electron Helper proxy on packaged builds (this PR's own `node.cmd`), this is OK in the happy path — but you should at least:

1. Document that the wrapper requires Electron-asar fs when only `gateway.asar` is present.
2. Or fall back through `gateway.asar.unpacked/node_modules/openclaw/package.json` (a real directory) as an additional candidate. Then `createRequire` from there at least gets you `@napi-rs/canvas` even without asar fs.

Related: the existing dynamic `import(pathToFileURL(resolved).href)` on an asar virtual path for the *ESM* `pdf.mjs` historically has edge cases in Electron's loader — please smoke-test this on an actual `ONECLAW_GATEWAY_ASAR=1` build before merging.

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude (general-purpose)

File: builtin-skills/pdf-extract/scripts/extract.mjs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  try {
+    ({ getDocument } = await importGatewayDependency("pdfjs-dist/legacy/build/pdf.mjs"));
+  } catch (err) {
+    fail(`pdfjs-dist not available: ${err.message}`);
+  }
+
+  // verbosity: 0 (ERRORS) silences pdfjs warnings on stderr (e.g. "Indexing
+  // all PDF objects", "standardFontDataUrl not configured"). The output-format
+  // contract reserves stderr for hard errors only.
+  const pdf = await getDocument({ data: new Uint8Array(buffer), disableWorker: true, verbosity: 0 }).promise;
```
Comment (oriengy) [inline builtin-skills/pdf-extract/scripts/extract.mjs]: **`verbosity: 0` does not silence one stray `console.log` in pdfjs — risk of breaking the one-line-stdout contract.**

pdfjs-dist 5.7.x has exactly one unguarded `console.log`:

```
build/pdf.mjs:1458
console.log("Deprecated API usage: " + details);
```

This is not gated by `verbosity` — if our code (or, more likely, a future pdfjs version) hits a deprecated API path, the message goes to **stdout** before the JSON line. That silently breaks the documented `output-format.md` contract ("Exactly one line of JSON terminated by `\n`. No banner, no progress, no trailing log lines.") and any naive `JSON.parse` consumer on the caller side fails with `Unexpected token D in JSON...`.

Recommended defence-in-depth at the top of `main()`:

```js
const origLog = console.log;
console.log = (...a) => process.stderr.write(`[pdf-extract pdfjs] ${a.join(" ")}\n`);
```

Or capture everything pre-`getDocument` and forward to stderr. Cheap and removes a real-world footgun.

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude (general-purpose)

File: builtin-skills/pdf-extract/scripts/extract.mjs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+function parsePages(spec) {
+  if (!spec) return null;
+  const out = [];
+  for (const part of spec.split(",")) {
+    const n = Number.parseInt(part.trim(), 10);
+    if (!Number.isFinite(n) || n < 1) fail(`invalid page number: ${part}`);
+    out.push(n);
+  }
```
Comment (oriengy) [inline builtin-skills/pdf-extract/scripts/extract.mjs:116]: **`parsePages` silently accepts garbage tokens.**

`Number.parseInt("3abc", 10) === 3`, so:

- `--pages 3abc,4xyz` parses as `[3, 4]` without an error.
- `--pages 1.5,2.9` parses as `[1, 2]`.
- `--pages ""` (empty after `split(",")` produces `[""]`) → `parseInt("", 10) === NaN` → falls into the `!Number.isFinite(n)` branch and fails with `invalid page number:` (note the trailing space) — message is technically correct but visually unhelpful.

If you want this script to be agent-friendly, prefer a stricter regex parser so `--pages 1.5` errors loudly rather than silently dropping the `.5`. Same goes for trailing/leading commas like `--pages 1,,3` which today parses as `[1]` then fails on the empty middle token.

Low severity, but worth tightening because the agent typically can't tell the difference between "page 3 had no text" and "I asked for page 3.5 and got page 3 instead".

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude (general-purpose)

File: builtin-skills/pdf-extract/scripts/extract.mjs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+  // verbosity: 0 (ERRORS) silences pdfjs warnings on stderr (e.g. "Indexing
+  // all PDF objects", "standardFontDataUrl not configured"). The output-format
+  // contract reserves stderr for hard errors only.
+  const pdf = await getDocument({ data: new Uint8Array(buffer), disableWorker: true, verbosity: 0 }).promise;
+  const pageCount = pdf.numPages;
+  const effectivePages = pageNumbers
+    ? pageNumbers.filter((p) => p >= 1 && p <= pageCount)
+    : Array.from({ length: pageCount }, (_, i) => i + 1);
```
Comment (oriengy) [inline builtin-skills/pdf-extract/scripts/extract.mjs]: **Empty `effectivePages` is silently swallowed and produces a misleading `fallbackImages: true` result.**

If the caller passes e.g. `--pages 99,100` on a 5-page PDF, the filter drops all entries → `effectivePages = []`. Then:

1. The text loop runs 0 times → `text = ""`.
2. `text.trim().length (= 0) < minTextChars (200)` → enters the image-fallback branch.
3. The image loop also runs 0 times → `imagePaths = []`.

Result: `{"text":"","fallbackImages":true,"imagePaths":[],"pageCount":5,"extractedPages":[],"truncatedText":false}` — which tells the model "this is a scanned PDF that produced no images", a complete fabrication.

Fix: short-circuit. Right after computing `effectivePages`, if it is empty AND `pageNumbers` was non-null and non-empty, `fail("requested pages out of range: ...")`. Otherwise emit a clearer `fallbackImages: false` text-empty result.

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude (general-purpose)

File: builtin-skills/pdf-extract/scripts/extract.mjs [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+      .filter(Boolean)
+      .join(" ");
+    if (pageText) textParts.push(pageText);
+  }
+  let text = textParts.join("\n\n");
+  let truncatedText = false;
+  if (text.length > DEFAULTS.maxChars) {
+    text = text.slice(0, DEFAULTS.maxChars);
+    truncatedText = true;
+  }
```
Comment (oriengy) [inline builtin-skills/pdf-extract/scripts/extract.mjs]: **Truncation can leave the text just below `minTextChars` and route a normal text PDF into the image-fallback path.**

Order of operations:

1. Join all per-page text.
2. Truncate joined text to `maxChars = 200_000`.
3. Check `text.trim().length >= 200`.

For a typical text PDF this is fine, but the truncation is unconditional — if you ever lower `maxChars` (e.g. to debug something at 100) you can easily produce a `<200` chars result and silently fall into rendering 200+ PNGs. The two budgets are coupled invisibly.

Cheap hardening: do the `minTextChars` threshold check on **untruncated** text (i.e. on `textParts.join(...).trim().length`) before clamping. That decouples the two budgets and matches user intent ("did this PDF actually have text content at all?" vs "how much text do we expose to the model?").

Not a blocker at the current 200k/200 values, but it's a foot-cannon hiding in the next config change.

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude (general-purpose)

File: builtin-skills/pdf-extract/scripts/extract.mjs [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+  await mkdir(outDir, { recursive: true });
+
+  const imagePaths = [];
+  const pixelBudget = Math.max(1, DEFAULTS.maxPixels);
+  for (const pageNum of effectivePages) {
+    const page = await pdf.getPage(pageNum);
+    const viewport = page.getViewport({ scale: 1 });
+    const pagePixels = viewport.width * viewport.height;
+    const scale = Math.min(1, Math.sqrt(pixelBudget / Math.max(1, pagePixels)));
+    const scaled = page.getViewport({ scale: Math.max(0.1, scale) });
```
Comment (oriengy) [inline builtin-skills/pdf-extract/scripts/extract.mjs]: **Pixel-budget math: the `Math.max(0.1, scale)` floor can blow past `maxPixels` for huge pages.**

Given a page of `pagePixels = 4×10⁸` (e.g. 20000×20000), `scale = sqrt(4e6 / 4e8) = 0.1` — exactly at the floor, fine. But for a 100000×100000 page (which exists in some CAD-exports-to-PDF), `scale = sqrt(4e6 / 1e10) ≈ 0.02`, clamped up to `0.1`. Resulting canvas: `10000 × 10000 = 1e8` pixels = 100 MB per page just for the RGBA backing buffer, then PNG encode, then a file write — likely OOM-kills the Helper process.

The gateway's `input-files-0oShoO1j.js` has the same code (per `reference/algorithm.md`), but the gateway also has a `maxPages` cap. This skill explicitly removes the page cap ("caller owns the risk"), so the per-page worst case is the only safety net left and it's broken by the `0.1` floor.

Options:

- Drop the `0.1` floor entirely and accept that pathological pages render at low resolution.
- Or bump the floor calculation to honour `maxPixels` strictly: `scale = Math.max(scaleFromPixels, MIN_USABLE_SCALE)` only when `scaleFromPixels * pageDim > MIN_USABLE_DIM`, else error out.
- Or add a hard upper bound on the canvas dimension (e.g. `Math.min(scaled.width, 8192)` — most PNG decoders cap there).

Given the PR's "caller owns risk" framing this may be acceptable, but please document the OOM mode in SKILL.md's Hard Limits table so it's not a surprise.

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude (general-purpose)

File: builtin-skills/pdf-extract/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+## Quick command
+
+```
+node <skill-dir>/scripts/extract.mjs <absolute-pdf-path>
+```
+
+`<skill-dir>` is the directory containing this `SKILL.md`. Optional flags:
+
+- `--pages 1,3,5` — extract specific pages (1-based). Default: all pages.
```
Comment (oriengy) [inline builtin-skills/pdf-extract/SKILL.md:37]: **Quick command does not mention `OPENCLAW_INSTALL_ROOT` — when the agent invokes it from `~/.openclaw/workspace/skills/pdf-extract/` outside the gateway's env, it will fail.**

The skill ships into `~/.openclaw/workspace/skills/pdf-extract/` on first launch. From there, normal Node module lookup walks `~/.openclaw/workspace/skills/pdf-extract/` → `~/.openclaw/workspace/skills/` → `~/.openclaw/workspace/` → `~/.openclaw/` → `~/` looking for `node_modules/pdfjs-dist`. None of those directories have one. So the agent depends 100% on `OPENCLAW_INSTALL_ROOT` being set in the env.

`OPENCLAW_INSTALL_ROOT` *is* set on the gateway subprocess (`src/gateway-process.ts:202`) and is normally inherited by anything the gateway spawns, but not every code path in OneClaw passes the gateway env down to the agent's Bash tool. A user testing this from a plain shell (e.g. for debugging, or via a different agent harness) will get `pdf-extract: pdfjs-dist not available: ...; OPENCLAW_INSTALL_ROOT is not set` and have no clue from SKILL.md what to do — they have to dig into `reference/runtime-discovery.md`.

Fix one of:

1. Surface this in the Quick command: `OPENCLAW_INSTALL_ROOT=... node <skill-dir>/scripts/extract.mjs <pdf>`, with a note that the gateway-spawned env normally sets it for you.
2. Or have the script self-discover the install root from a sibling fact (e.g. a `.openclaw-install-root` file written next to the workspace skills dir at install time).

The failure mode of "agent reads SKILL.md, copies the Quick command verbatim, gets a cryptic error" is the most likely real-world breakage of this skill.

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude (general-purpose)

File: builtin-skills/pdf-extract/examples/invocations.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+```
+
+Because page text falls below 200 chars, the script renders each page to a
+PNG and returns the file paths:
+
+```json
+{
+  "text": "",
+  "fallbackImages": true,
+  "imagePaths": [
```
Comment (oriengy) [inline builtin-skills/pdf-extract/examples/invocations.md:53]: **Drift between docs and script: scanned-PDF example shows `"text":""`, but the script emits whatever partial text was extracted (could be 100+ chars).**

`extract.mjs` populates `text` from `textParts.join("\n\n")` *before* the threshold check, then enters the fallback branch on `text.trim().length < minTextChars`, and writes the same `text` value into the fallback JSON (`scripts/extract.mjs:166`). It does **not** reset `text` to `""`.

So this documented example is incorrect — a real scanned PDF with even partial OCR text on the cover page will produce something like:

```json
{"text":"Invoice #1234 Date: ...","fallbackImages":true,"imagePaths":[...]}
```

Not a bug per se (the schema in `reference/output-format.md` says `text` *may* be `""`), but the example here misleads anyone consuming the output. Either:

- Update this example to show a realistic non-empty `text` value, or
- Have the script `text = ""` in the fallback branch so docs and reality agree.

My mild preference is the latter: when `fallbackImages: true`, leave text-mining to the image-reading model rather than handing it a sentence-fragment that pretends to be the document.

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude (general-purpose)

File: scripts/afterPack.js [hunk 9]
```diff
@@ -288,6 +296,24 @@ function buildWindowsElectronProxyScript(productName, cliEntryPath) {
...
+    "@echo off",
+    'set "ELECTRON_RUN_AS_NODE=1"',
+    `set "APP_EXE=${mainExe}"`,
+    `set "APP_HELPER=${helperExe}"`,
+    'if exist "%APP_HELPER%" (',
+    '  "%APP_HELPER%" %*',
+    ") else (",
+    '  "%APP_EXE%" %*',
+    ")",
+  ].join("\r\n") + "\r\n";
```
Comment (oriengy) [inline scripts/afterPack.js:314]: **Missing `setlocal` / explicit exit-code propagation in `node.cmd`.**

Two small but real Windows-batch hygiene problems compared to `buildWindowsElectronProxyScript` (which has the same issue, but already in use is one thing — copying the pattern into a *more frequently invoked* `node` proxy makes it worse):

1. **No `setlocal`** — `set "ELECTRON_RUN_AS_NODE=1"`, `APP_EXE`, `APP_HELPER` all leak into the caller's environment when this proxy is `call`-invoked from another `.cmd` script (e.g. when something runs `call node extract.mjs` inside a wrapper). The next command in that parent script sees `ELECTRON_RUN_AS_NODE=1` even though it has nothing to do with Electron.
2. **No explicit `exit /b %ERRORLEVEL%`** — cmd.exe usually carries the last command's ERRORLEVEL out implicitly, but if the user happens to invoke this via `cmd /c node.cmd ...` and pipes / redirections muddy the chain, the exit code is occasionally lost. Explicit is safer.

Fix:

```cmd
@echo off
setlocal
set "ELECTRON_RUN_AS_NODE=1"
set "APP_HELPER=%~dp0..\..\..\OneClaw Helper.exe"
set "APP_EXE=%~dp0..\..\..\OneClaw.exe"
if exist "%APP_HELPER%" (
  "%APP_HELPER%" %*
) else (
  "%APP_EXE%" %*
)
exit /b %ERRORLEVEL%
```

The same fix would apply to `buildWindowsElectronProxyScript`, but that's pre-existing and out of scope; just don't propagate the bug pattern into the new `node` proxy.

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude (general-purpose)

File: scripts/afterPack.test.js [hunk 10]
```diff
@@ -39,6 +39,14 @@ test("Windows afterPack wrapper 应优先调用 Helper.exe 并回退主 exe", ()
   assert.match(npmCmd, /if exist/i);
   assert.match(npmCmd, /OneClaw\.exe/);
 
+  // node.cmd 代理：与 macOS runtime/node 对称，原样转发 %*，不绑定 CLI 入口
+  assert.equal(fs.existsSync(path.join(runtimeDir, "node.cmd")), true);
+  const nodeCmd = fs.readFileSync(path.join(runtimeDir, "node.cmd"), "utf-8");
+  assert.match(nodeCmd, /ELECTRON_RUN_AS_NODE=1/);
+  assert.match(nodeCmd, /OneClaw Helper\.exe/);
+  assert.match(nodeCmd, /%\*/);
+  assert.doesNotMatch(nodeCmd, /npm-cli\.js|npx-cli\.js/);
```
Comment (oriengy) [inline scripts/afterPack.test.js:48]: **Test coverage is shallow — none of the load-bearing claims of the `node.cmd` proxy are actually exercised.**

The new assertions only check that the file exists, contains the literal strings `ELECTRON_RUN_AS_NODE=1`, `OneClaw Helper.exe`, `%*`, and doesn't contain `npm-cli.js|npx-cli.js`. None of these would catch the most likely regressions:

1. **Path arithmetic**: a typo like `%~dp0..\..\<Product>.exe` (one `..` fewer) would still match all the regexes above. Add an assertion on the literal `%~dp0..\..\..\OneClaw Helper.exe` chunk so the up-three-levels assumption is locked in.
2. **Helper-then-fallback order**: assert the helper branch appears *before* the fallback branch (regex `OneClaw Helper\.exe[\s\S]*else[\s\S]*OneClaw\.exe`). Today the test would pass even if `if exist` checked `OneClaw.exe` first.
3. **`%*` is used in both branches**: assert *two* occurrences of `%*`, not one. The current `assert.match(nodeCmd, /%\*/)` passes even if only the fallback branch forwards args.
4. **No `setlocal` leak** (see my comment on `afterPack.js:314`): once you add `setlocal`/`exit /b`, add asserts for them so future refactors don't silently regress.

The second test (`保留展开的 gateway node_modules`) adds `mirrorDir` / `webbridgeDir` fixtures that look unrelated to this PR's stated scope — looks like it was needed to make the full `afterPack` flow run end-to-end after some other change, but the diff doesn't add any assertions on those new fixtures' destination. Either add an assertion that the mirror/CRX got copied, or leave a comment explaining why these are required test scaffolding rather than the unit-under-test.

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude (general-purpose)

File: builtin-skills/pdf-extract/scripts/extract.mjs [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+async function importGatewayDependency(specifier) {
+  try {
+    return await import(specifier);
+  } catch (directErr) {
+    for (const packageJson of bundledOpenclawPackageJsons()) {
+      if (!existsSync(packageJson)) continue;
+      try {
+        const resolved = createRequire(packageJson).resolve(specifier);
+        return await import(pathToFileURL(resolved).href);
+      } catch {}
```
Comment (oriengy) [inline builtin-skills/pdf-extract/scripts/extract.mjs:78]: **Catch-all `catch {}` swallows useful diagnostics.**

If `createRequire(packageJson).resolve(specifier)` or the dynamic `import()` of the resolved file fails, the error is dropped on the floor and the loop moves to the next candidate. When all candidates fail, the user only sees the *outer* `directErr.message` (the failure of the first plain `await import(specifier)`), which doesn't tell them that e.g. "resolved `pdfjs-dist` inside `gateway.asar`, but ESM loader couldn't read it because asar isn't mounted."

At least collect the per-candidate errors and append them to the thrown message — or log them to stderr — so an operator debugging a packaging regression has a fighting chance.

```js
const errors = [`direct: ${directErr.message}`];
for (const packageJson of bundledOpenclawPackageJsons()) {
  if (!existsSync(packageJson)) continue;
  try {
    const resolved = createRequire(packageJson).resolve(specifier);
    return await import(pathToFileURL(resolved).href);
  } catch (e) {
    errors.push(`${packageJson}: ${e.message}`);
  }
}
throw new Error(errors.join("; "));
```

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude (general-purpose)

File: builtin-skills/pdf-extract/scripts/extract.mjs [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    imagePaths.push(out);
+  }
+
+  process.stdout.write(JSON.stringify({
+    text, fallbackImages: true, imagePaths,
+    pageCount, extractedPages: effectivePages, truncatedText,
+  }) + "\n");
+}
+
+main().catch((err) => fail(err?.stack || String(err)));
```
Comment (oriengy) [inline builtin-skills/pdf-extract/scripts/extract.mjs]: **`main().catch((err) => fail(err?.stack || String(err)))` will write the full stack with embedded newlines to stderr — that's fine for stderr — but `pdf.getDocument(...)` and `page.render(...)` can throw `PasswordException` / `MissingPDFException` / `InvalidPDFException` from pdfjs. None of those are mapped to a useful user-facing message.**

An encrypted PDF currently emits something like:

```
pdf-extract: PasswordException: No password given
    at PDFDocumentLoadingTask._getDocument (...pdfjs.../pdf.mjs:14831:9)
    ...
```

Which the agent then has to parse. Worth a small `try/catch` around `getDocument(...).promise` translating known pdfjs exception names into a stable, documented stderr message (`pdf-extract: encrypted pdf, password required`, `pdf-extract: corrupt pdf: <reason>`, etc.).

This would also let `reference/output-format.md`'s exit-code table grow a few additional codes if desired (e.g. exit 2 for "protected", exit 3 for "corrupt") — but even keeping the binary 0/1 contract, the stable stderr text is the part that matters for the calling model.

🤖 Generated with [Claude Code](https://claude.com/claude-code) — Claude (general-purpose)

File: builtin-skills/pdf-extract/scripts/extract.mjs [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+
+function fail(msg, code = 1) {
+  process.stderr.write(`pdf-extract: ${msg}\n`);
+  process.exit(code);
+}
+
+function bundledOpenclawPackageJsons() {
+  const installRoot = process.env.OPENCLAW_INSTALL_ROOT;
+  if (!installRoot) return [];
+  return ["gateway", "gateway.asar"].map((gatewayRoot) =>
```
Comment (oriengy) [inline builtin-skills/pdf-extract/scripts/extract.mjs]: **Fallback precedence is inverted relative to how OneClaw actually resolves the gateway.**

This probes `gateway/` before `gateway.asar`, while the app runtime resolves the packaged gateway by preferring `gateway.asar` when it exists. If a target/app output ever contains stale loose files next to a fresh ASAR, the copied workspace skill can load `pdfjs-dist` / `@napi-rs/canvas` from the wrong gateway tree and silently diverge from what OneClaw actually runs.

Smallest fix: mirror `resolveGatewayRoot()` precedence — try `gateway.asar` first, or skip `gateway/` whenever `gateway.asar` is present.

*Complementary to @Claude's `:38` finding (which flags that `gateway.asar` is a file and `existsSync(.../gateway.asar/node_modules/...)` only works under Electron's patched fs). Different angle on the same fragile area.*

🤖 Generated with [Codex via Claude Code](https://claude.com/claude-code) — Codex

File: builtin-skills/pdf-extract/scripts/extract.mjs [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+      : "OPENCLAW_INSTALL_ROOT is not set";
+    throw new Error(`${directErr.message}; ${hint}`);
+  }
+}
+
+function parseArgs(argv) {
+  const args = { pdfPath: null, pages: null, outDir: null };
+  for (let i = 0; i < argv.length; i++) {
+    const a = argv[i];
+    if (a === "--pages") args.pages = argv[++i];
```
Comment (oriengy) [inline builtin-skills/pdf-extract/scripts/extract.mjs]: **Bare `--pages` (no value) silently degrades to "extract every page".**

A bare `--pages` is accepted as if no page filter was provided because `argv[++i]` becomes `undefined`, then `parsePages(undefined)` returns `null`. That turns a malformed narrow-extraction request into "extract every page" — which is the risky path this flag is meant to **avoid** on large or scanned PDFs (long runtime, huge fallback PNG volume, context-window blowup).

Smallest fix: fail immediately when `--pages` has no following value, or when the next token starts with `--`.

🤖 Generated with [Codex via Claude Code](https://claude.com/claude-code) — Codex

File: builtin-skills/pdf-extract/scripts/extract.mjs [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+  let createCanvas;
+  try {
+    ({ createCanvas } = await importGatewayDependency("@napi-rs/canvas"));
+  } catch (err) {
+    fail(`@napi-rs/canvas not available for scanned-PDF fallback: ${err.message}`);
+  }
+
+  const outDir = args.outDir
+    ? path.resolve(args.outDir)
+    : path.join(tmpdir(), `oneclaw-pdf-${createHash("sha1").update(pdfPath).digest("hex").slice(0, 12)}`);
```
Comment (oriengy) [inline builtin-skills/pdf-extract/scripts/extract.mjs]: **Default fallback out-dir is deterministic from the PDF path — concurrent or repeated runs can return stale / mixed PNGs.**

The default output directory is derived only from the PDF path, so repeated or concurrent fallback renders for the same path write the same `page-N.png` files. If the file contents change, or two extractions run at once with different `--pages`, the returned `imagePaths` can point at stale or mixed images from another run — and the caller has no way to tell.

Smallest fix: use `fs.mkdtemp()` (which gives a guaranteed-unique suffix) or include a per-run nonce / file-content hash in the directory name.

🤖 Generated with [Codex via Claude Code](https://claude.com/claude-code) — Codex

**Comment by oriengy** [comment]
## Re-review @ head `4f029a2` — dual-agent (Claude + Codex)

Independent deep re-review after the `harden review fixes` commit. Two agents traced the full `extract.mjs` data flow (arg parsing → runtime/pdfjs discovery → page selection → text extraction → image fallback → JSON contract → error handling) independently; findings reconciled and verified against source below.

### Verdict: **merge-with-followup**
The hardening commit is solid. **All 8 issues from the previous round are genuinely fixed** (verified, including a live run of the text + fallback paths). One substantive issue remains open (already has an inline thread) and one architectural path warrants a packaged-build smoke test.

### Prior findings — all FIXED
| Issue | Status | Evidence |
|---|---|---|
| ASAR path precedence / file-vs-dir / `existsSync` | FIXED | `bundledOpenclawPackageJsons` now `["gateway.asar","gateway"]`; `asarRuntimeHint` keys off the asar **file's** own existence (L41-54) |
| pdfjs `console.log` → stdout JSON pollution | FIXED | `installQuietOutput` (L111) installed before pdfjs/canvas import (L256); final JSON via saved writer (L259) |
| `parseInt("3abc")` | FIXED | strict `/^[1-9]\d*(,…)*$/` + `Number.isSafeInteger` (L101-106) |
| bare `--pages` → extract-all | FIXED | `readFlagValue` hard-fails missing value (L74-77) |
| `Math.max(0.1,scale)` pixel-cap bypass | FIXED | `fitCanvasToPixelBudget` has no lower clamp, floor-then-decrement loop (L144-172) |
| deterministic PNG out-dir collision | FIXED | `uniqueOutputDir` = pid + `randomUUID()` (L138-141) |
| SKILL.md missing `OPENCLAW_INSTALL_ROOT` | FIXED | SKILL.md L24-32 |
| stderr stack-trace leak | FIXED | catch emits `err.message` only (L261, L266) |

### Still open (re-affirming the existing inline thread — NOT fixed by 4f029a2)
- **Out-of-range `--pages` → silent empty "success"** (`extract.mjs:188-190`). `--pages 999` on a 2-page PDF → `effectivePages=[]` → empty text → fallback path → emits `{"text":"","fallbackImages":true,"imagePaths":[],"extractedPages":[]}` with **exit 0**. The agent gets a "success" with zero content and no signal its page selection was wrong. Both agents independently re-found this (it's the live thread at L190). Fix: after filtering, `if (pageNumbers && effectivePages.length === 0) fail("no requested pages within document (pageCount=N)")`.

### New followup — verify before relying on ASAR builds
- **ASAR/ESM dependency `import()` is unverified on a packaged build** (`extract.mjs:64, 178`). `import(pathToFileURL(resolved).href)` for a `gateway.asar/node_modules/...` path relies on the skill being executed by an ASAR-capable runtime (Electron-as-node with `ELECTRON_RUN_AS_NODE=1`) **and** Electron 40's ESM loader honoring asar URLs. Production OneClaw ships ASAR (`gateway.asar.unpacked` present). The CommonJS asar patch is well-established; **ESM-over-asar is not exercised here**. Recommend a one-time smoke test of the fallback (scanned-PDF) path on an `ONECLAW_GATEWAY_ASAR=1` build. A failure degrades to the clear `asarRuntimeHint`, not a crash — so this is a verification item, not a blocker.

### ⚠️ Codex false positive — recorded so nobody acts on it
Codex (gpt-5.5, xhigh effort) raised a **BLOCKER** at `extract.mjs:232`, claiming `page.render({ canvas, viewport })` is broken because pdfjs requires a 2D `canvasContext`. **This is incorrect for the bundled `pdfjs-dist@5.7.284`** (openclaw's pinned version). Verified against the actual package source:
- `RenderParameters.canvas` is the **recommended** parameter; `canvasContext` is documented as "for backwards compatibility … it is recommended to use the `canvas` parameter instead" (`types/src/display/api.d.ts:403-422`).
- Implementation: `render({ canvasContext, canvas = canvasContext.canvas, … })` (`legacy/build/pdf.mjs:21930`); `this._canvasContext = params.canvas ? null : params.canvasContext` (L23137); `const canvasContext = this._canvasContext || this._canvas.getContext("2d", …)` (L23171) — **pdfjs derives the 2D context from the passed `canvas` itself.**

The current `{ canvas, viewport }` call is the correct pdfjs v5 API (Codex appears to have reasoned from the pre-v5 API where `canvasContext` was required). No change needed.

---
*Re-review by Claude (general-purpose) + Codex (gpt-5.5, xhigh), reconciled and verified against `pdfjs-dist@5.7.284` source.*
